test/behaviour/connection/ConnectionStepsBase.java

```diff
@@ -100,4 +93,15 @@ public synchronized void after() {
         client = null;
         System.out.println("ConnectionSteps.after");
     }
+
+    abstract GraknClient createGraknClient(String address);
+
+    void connectionHasBeenOpenedImpl() {
+        assertNotNull(client);
+        assertTrue(client.isOpen());
+    }
+
+    void connectionDoesNotHaveAnyDatabaseImpl() {
+        assertTrue(client.databases().all().isEmpty());
+    }
```

Comment:
Having to suffix methods with `impl` is also ugly. Does it work if we rename `beforeImpl` and `afterImpl` to `before` and `after` and call `super.before()` in the subclasses?

Moreover, the step `connection does not have any database` should be moved to `DatabaseSteps`. I'm pretty sure nothing will break because _every_ test file uses `DatabaseSteps`.

As for `connection has been opened`, again, can we rename it to match the method in the subclass and call `super.connection_has_been_opened()`?

(Incidentally, did you test simply defining the test steps (`@When("connection has been opened")`) in this file? I assume you tried and Cucumber didn't like it)

Reply:
> Having to suffix methods with impl is also ugly. Does it work if we rename beforeImpl and afterImpl to before and after and call super.before() in the subclasses?

> Moreover, the step connection does not have any database should be moved to DatabaseSteps. I'm pretty sure nothing will break because every test file uses DatabaseSteps.

> As for connection has been opened, again, can we rename it to match the method in the subclass and call super.connection_has_been_opened()?

all done in https://github.com/graknlabs/client-java/pull/251/commits/82e0af303bbd22fbd370e3537fa2aea7235e617b

Reply:
> (Incidentally, did you test simply defining the test steps (@When("connection has been opened")) in this file? I assume you tried and Cucumber didn't like it)

Yes, this is correct. I tried it and it won't work